### DISCUSSION TITLE
Provide "hook" after file generation

### DISCUSSION THREAD

**Discussion by svenjacobs** [ivanhofer/typesafe-i18n#185] [Ideas]
First of all thanks for this great library! 👏🏼 

It would be nice to be able to configure a "hook" or command which is executed after `typesafe-i18n` has created / updated the generated files. For example I would like to format these files with Prettier according to my projects code formatting rules.

The hook / command could be configured in `.typesafe-i18n.json` for instance.

**Comment by ivanhofer** [comment]
@svenjacobs sounds like a gread addition!
How would you expect to configure this?

Like this where you specify the command that then get's executed 1:1 in a [`childProcess`](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback)?
```json
{
   "runOnGeneratorEnd": "npm run prettier"
}
```

**Reply by svenjacobs** [reply]
Yes, this would be absolutely fine for my use case 👍🏼

**Comment by ivanhofer** [comment]
implemented in version `4.2.0`

find the docs [here](https://github.com/ivanhofer/typesafe-i18n#runaftergenerator)

**Reply by svenjacobs** [reply]
Awesome! Thanks for this super quick implementation 😃

**Reply by svenjacobs** [reply]
I just tried it out but I have a weird behaviour:

It seems that **sometimes**, although I see that `typesafe-i18n` is calling `prettier` and Prettier prints output to the terminal, the Prettier changes are not written to the files. I also noticed that `typesafe-i18n` is detecting changes to generated files like `i18n-types.ts` (for example when I manually run Prettier) and is calling the file generation process again with Prettier at the end. Could this be a race condition? When Prettier is run as a child process, `typesafe-i18n` detects changes to the files and writes them again, overwriting the Prettier output?

I have this in my `package.json`:

```json
"scripts": {
  "dev": "run-p dev:kit typesafe-i18n",
  "dev:kit": "svelte-kit dev",
  "build": "svelte-kit build",
  "package": "svelte-kit package",
  "preview": "svelte-kit preview",
  "check": "svelte-check --tsconfig ./tsconfig.json",
  "check:watch": "svelte-check --tsconfig ./tsconfig.json --watch",
  "lint": "prettier --ignore-path .gitignore --check --plugin-search-dir=. . && eslint --ignore-path .gitignore .",
  "format": "prettier --ignore-path .gitignore --write --plugin-search-dir=. .",
  "format:i18n": "prettier --ignore-path .gitignore --write --plugin-search-dir=. ./src/i18n",
  "typesafe-i18n": "typesafe-i18n"
},
```

I usually call `npm run dev` while developing.

My `.typesafe-i18n.json`

```json
{
  "$schema": "https://unpkg.com/typesafe-i18n@4.2.0/schema/typesafe-i18n.json",
  "adapter": "svelte",
  "runAfterGenerator": "npm run format:i18n"
}
```

**Reply by ivanhofer** [reply]
There is a 1 second delay between running the hook and then reacting to changes inside `src/i18n` again.
Try to increase the timeout in `node_modules/typesafe-i18n/cli/typesafe-i18n.mjs` and let me know if it works

![image](https://user-images.githubusercontent.com/21335119/155093134-24fbc7e7-556e-458e-8b6b-631e4e21b61f.png)

**Reply by svenjacobs** [reply]
I increased the delay to 5 seconds but it is still randomly working or not. Instead of a timeout, couldn't you wait until the child process is terminated? I believe this is much safer, since we cannot assume how long Prettier (or any other command that is executed via `runAfterGenerator`) takes on different systems and many files?

But I have the feeling that the issue is not `disableChangeDetection` here...

**Reply by ivanhofer** [reply]
when this timeout gets executed, the `prettier` command is already completed for 1 second. I tested it and even 100ms worked for me.
Can you create a repo so I can test it with my setup?

**Reply by svenjacobs** [reply]
What I was trying to say is that a `setTimeout` looks like a possible source of problems, when for instance the command takes longer than the defined timeout, regardless whether the current problem is related to the timeout or not 😉 Couldn't you use [execSync](https://nodejs.org/api/child_process.html#child_processexecsynccommand-options)?

Anyway, I will try to create a sample project that reproduces this problem.

**Reply by svenjacobs** [reply]
I just see that you **are** using `execSync`. So why is the timeout needed anyway?

**Reply by ivanhofer** [reply]
i assume the file watcher has a delay of a few ms until changes are detected

**Reply by svenjacobs** [reply]
Please see [this repository](https://github.com/svenjacobs/typesafe-i18n-prettier-example). I created the SvelteKit sample project with `npm init svelte@next` and afterwards added `typesafe-i18n` via `npx typesafe-i18n --setup-auto`.

When you run `npm run dev` you will see that the generated files are changed and not formatted according to Prettier rules. When you keep changing translations you will see that it sometimes works, other times not.

I'm using Node `17.5.0` by the way.

**Reply by ivanhofer** [reply]
#itWorksOnMyMachine 😅
I also tried it with a timeout of 10ms and there it works too.
Which operating system are you using? Is it a low-spec device you are using?

**Reply by svenjacobs** [reply]
This is a 2019 MacBook Pro with a 2,6 GHz 6-Core Intel Core i7 on macOS Monterey 12.2.1.

Do you think it has something to do with the timeout? I tried it with 5 seconds, Prettier was long finished according to the output, and I still had this issue. Maybe it's a problem with properly flushing file changes to the file system from a child process or something like that?

**Reply by ivanhofer** [reply]
I'm using `chokidar` to watch for changes made to the `src/i18n` directory.
Can you insert a console log here to see when `chokidar` detects changes and what the value of `disableChangeDetection` is.
![image](https://user-images.githubusercontent.com/21335119/155274240-08a9f94b-a98b-4747-b0d4-b086df87675c.png)
Maybe also insert a log when the timeout finishes?

**Reply by svenjacobs** [reply]
This is the output:

```
➜  app git:(main) g status
On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
➜  app git:(main) npm run typesafe-i18n

> app@0.0.1 typesafe-i18n
> typesafe-i18n

[typesafe-i18n] version 4.2.0
[typesafe-i18n] generating files for TypeScript version: '4.5.x'
[typesafe-i18n] options: { adapter: 'svelte', runAfterGenerator: 'npm run format:i18n' }
[typesafe-i18n] watcher started in: './src/i18n/'
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = false
[typesafe-i18n] generated file: './src/i18n//i18n-types.ts'
[typesafe-i18n] generated file: './src/i18n//i18n-util'
[typesafe-i18n] generated file: './src/i18n//i18n-util.sync'
[typesafe-i18n] generated file: './src/i18n//i18n-util.async'
[typesafe-i18n] generated file: './src/i18n//i18n-svelte'
[typesafe-i18n] running command 'npm run format:i18n'
[typesafe-i18n] output:
>
>
> > app@0.0.1 format:i18n
> > prettier --write --plugin-search-dir=. ./src/i18n
>
> src/i18n/de/index.ts 28ms
> src/i18n/en/index.ts 7ms
> src/i18n/formatters.ts 8ms
> src/i18n/i18n-svelte.ts 1ms
> src/i18n/i18n-types.ts 28ms
> src/i18n/i18n-util.async.ts 1ms
> src/i18n/i18n-util.sync.ts 1ms
> src/i18n/i18n-util.ts 1ms
>
[typesafe-i18n] ... all files are up to date
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! CHOKIDAR disableChangeDetection = true
[typesafe-i18n] !!! TIMEOUT
^C
➜  app git:(main) ✗ g status
On branch main
Your branch is ahead of 'origin/main' by 1 commit.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   src/i18n/i18n-svelte.ts
	modified:   src/i18n/i18n-util.async.ts
	modified:   src/i18n/i18n-util.sync.ts
	modified:   src/i18n/i18n-util.ts
```

**Reply by svenjacobs** [reply]
As you can see, the generated files have changed after Prettier because the projects code formatting was not applied. Interestingly, the translation files `src/i18n/de/index.ts` and `src/i18n/en/index.ts` where not changed. Are all file streams properly closed / flushed **before** the command is executed?

**Reply by svenjacobs** [reply]
By the way, what system do you use? Is it a different one than mine, maybe Linux or Windows? It could explain different behaviour of file system operations.

**Reply by ivanhofer** [reply]
@svenjacobs looks normal to me. After the `!!! TIMEOUT` comment, nothing should happen. The generator only runs when `disableChangeDetection` is set to `false`. I think you have aborted the process to early.

> even if multiple `chokidar` changes are detected, they are beeing debounced to run the generator only once

`typesafe-i18n` doesn't touch the dictionary files. It only generates the files in the root of `src/i18n`

I think everything get's flushed, but I can take a deeper look at it tomorrow.

**Reply by ivanhofer** [reply]
I'm running WSL on windows

**Reply by svenjacobs** [reply]
> `typesafe-i18n` doesn't touch the dictionary files. It only generates the files in the root of `src/i18n`

I know, but Prettier also touched the dictionary files and they were the only ones that were not modified afterwards. So only the generated files were formatted wrongly. I have a strong feeling that it has something to do with how the generated files are written. As I said, maybe streams are not flushed before the command is executed?

**Reply by svenjacobs** [reply]
> I think you have aborted the process to early.

No, I kept it running for a few more seconds and there was no further output.

**Reply by svenjacobs** [reply]
~~I'm not familiar with `esbuild` but it seems that the `build()` function returns a `Promise`. To be safe, I guess you should wait [here](https://github.com/ivanhofer/typesafe-i18n/blob/c2884e6f3143de7970bc95490adca4851e634381/packages/adapter-svelte/esbuild.ts#L15) on the returned Promises. So instead of `forEach` use `map`. For example (simplified pseudo code):~~

```javascript
const promises = formats.map((fmt) => build(...))
await Promise.all(promises)
```

**Reply by svenjacobs** [reply]
Sorry, I believe the linked file is the wrong location to look for. Still getting used to the project 😅

**Reply by svenjacobs** [reply]
I'm not sure but maybe [here](https://github.com/ivanhofer/typesafe-i18n/blob/c2884e6f3143de7970bc95490adca4851e634381/packages/generator/src/file-utils.ts#L80) an `await` is missing? I'm currently on mobile and don't have an IDE at hand.

**Reply by ivanhofer** [reply]
good catch. I added the await in version `4.2.2`. But I still don't think this is the issue

**Reply by svenjacobs** [reply]
Yes, that seems to be the problem. I just changed line 21430 of `node_modules/typesafe-i18n/cli/typesafe-i18n.mjs` (in function `writeNewFile()`) to

```javascript
await writeFile(getFileName(path4, file), content);
```

and the problem is gone! 🎉

**Reply by ivanhofer** [reply]
oh really? 🙈🙈
but how? I still don't unedrstand how this would re-trigger the generator 🤔

**Reply by svenjacobs** [reply]
> oh really? 🙈🙈
> but how? I still don't unedrstand how this would re-trigger the generator 🤔

I don't think it has anything to do with retriggering the generator. I believe because of the missing `await`, the changes of the generator were sometimes flushed to the file after Prettier was executed. It's a classic race condition, since it was working sometimes.

**Reply by ivanhofer** [reply]
never mind, thanks for helping identifying the bug :)

**Reply by svenjacobs** [reply]
You're welcome 🙂

**Reply by svenjacobs** [reply]
Hey, it's me again, unfortunately 😉 

It seems I'm now actually running into the timeout issue. Sometimes after Prettier finished, changes are detected again which causes regeneration of files which in turn executes Prettier again. This sometimes happens two, three times in a row.

When increasing the timeout to 2000ms in `cli/typesafe-i18n.mjs` this does not happen.

Instead of a timeout, couldn't you use a boolean flag? Something like `skipNextChangeDetection` which is set to `true` in `runCommandAfterGenerator()` before the `execSync()` call and which is set to `false` in the Chokidar callback?

**Reply by ivanhofer** [reply]
Sorry to hear that.

That's exactly what I do ^^ + the additional 1000ms timeout to account for possible `chokidar` hickups.

1. changes are detected
2. 100ms waiting if another change was triggered
3. disable  change detection
4. start generation of files
5. run custom `runAfterGenerator` command
6. wait 1000ms
7. enable change detection

That's also why it seemd odd to me that the solution we found before was actually working until now.

I found a [place in the code](https://github.com/ivanhofer/typesafe-i18n/commit/0096dc6536121f51560d2dad75759af1e85568e3#diff-c7880ce51f904f77ed5479d69229ff5b9f07883bb10dc263250b1bec2f0faebcR52) that could be an issue.
Please try it with version `4.3.4` and let me know if it helps. But I don't think this will be the correct solution.

Maybe I should try to replace `chokidar` with something more "modern" and lightweight like `https://github.com/lukeed/watchlist`

**Reply by svenjacobs** [reply]
It seems that the latest fix has solved this problem. Thank you! 😃
